### PR TITLE
Scene Import

### LINKED ISSUE #714: Scene imports - Ability to "import" and modify another scene file inside tangram scene yaml
Refer JS PR and conversation in the PR for all conceptual details: https://github.com/tangrams/tangram/pull/278

### PR DESCRIPTION
**Pull request by tallytalwar**
For conceptual description refer JS PR: https://github.com/tangrams/tangram/pull/278

Fixes #714

TODO:
- [x] Maintain relative paths on imported assets, example normalize texture paths with respect to the import scene file, etc
-  Network request for imported scenes and handle downloading of all the referenced assets in the scene. (to be done in a separate PR, as this will require more work on the networking side of things) Refer: https://github.com/tangrams/tangram-es/pull/776
- [x] Unit tests

### CODE REVIEW

**Comment by tallytalwar** [comment]
@karimnaaji you want to have a look at this?

The only thing left for the basic import (non network) is to add unit tests.

**Comment by karimnaaji** [comment]
Sure, I will start to look at this today.

File: core/src/scene/importer.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void Importer::setNormalizedTexture(Node& texture, const std::vector<std::string>& names,
+        const std::string& parentPath) {
+
+    for (auto& name : names) {
+        if (m_textureNames.find(name) == m_textureNames.end()) {
+            m_textureNames.insert(name);
+            if (names.size() > 1) {
+                texture.push_back(normalizePath(name, parentPath));
```
Comment (karimnaaji) [inline core/src/scene/importer.cpp]: It should not `push_back` on the container here, we should access `texture[current_index]` and update it instead.

File: core/src/scene/importer.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                }
+            }
+        }
+    }
+}
+
+bool Importer::loadScene(const std::string& path) {
+
+    auto scenePath = path;
+    auto sceneString = stringFromFile(setResourceRoot(scenePath.c_str()).c_str(), PathType::resource);
```
Comment (karimnaaji) [inline core/src/scene/importer.cpp]: `s_resourceRoot` is static within the platform so once modified we would stick with that modified resource root for any other loaded resource. I guess that `setResourceRoot` should be called again when going out of the recursion, otherwise it would consider the root used for the last imported scene (which may be a different/deeper path).

**Comment by karimnaaji** [comment]
@tallytalwar , this looks good to me! Last thing would be unit tests for some edge cases and clarifying the comments above.

File: core/src/scene/importer.cpp (hunk 2 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Yes thats a good catch. Instead of calling it again, I can just call it once from `tangram.cpp` and pass just the resultant `scenePath` to the `Importer::loadScene`.

File: core/src/scene/importer.cpp (hunk 1 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Yup, this is a bug, would have added more textures to the YAML::Node. Will fix this.

Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Fixed by 6563ea1

File: core/src/scene/importer.cpp (hunk 2 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Fixed by 6563ea1

File: core/src/scene/importer.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        }
+    }
+}
+
+void Importer::setNormalizedTexture(Node& texture, const std::vector<std::string>& names,
+        const std::string& parentPath) {
+
+    for (size_t index = 0; index < names.size(); index++) {
+        auto& name = names[index];
+        if (m_textureNames.find(name) == m_textureNames.end()) {
```
Comment (karimnaaji) [inline core/src/scene/importer.cpp]: I don't think that we need that test. If we have two scenes that reference the same texture, this will fail to normalize the texture path where one imports the other, we will fail in normalizing the path and the second path will try to load a resource that is not relevant to its own resource root.
In general I think there's also an issue with the texture management within the scene, where they should be stored by URI at some point. I will create an issue and look into fixing that texture management.

**Comment by karimnaaji** [comment]
https://github.com/tangrams/tangram-es/commit/e7c3b696cb6d9da4ba19d2f45f1794a6163aa690 looks good!

File: core/src/scene/importer.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    }
+}
+
+bool Importer::loadScene(const std::string& path) {
+
+    auto scenePath = path;
+    auto sceneString = stringFromFile(scenePath.c_str(), PathType::resource);
+    auto sceneName = getFilename(scenePath);
+
+    if (m_scenes.find(sceneName) != m_scenes.end()) { return true; }
```
Comment (blair1618) [inline core/src/scene/importer.cpp]: Differentiating scenes by their file name alone seems inadequate. If a scene imports both `folder_a/scene.yaml` and `folder_b/scene.yaml` then by this logic, I think the second will be skipped. 

Also, if this is meant to be an early exit from the function, it should happen before the expensive call to `stringFromFile()`.

Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Sure both are legitimate concerns, will fix these.

File: core/src/scene/importer.cpp [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+std::vector<std::string> Importer::getImportOrder() {
+
+    std::vector<std::pair<std::string, std::string>> dependencies;
+
+    for (const auto& scene : m_scenes) {
+        const auto& name = scene.first.k;
+        const auto& sceneRoot = scene.second;
+        for (const auto& import : getScenesToImport(sceneRoot)) {
+            auto importName = getFilename(import);
+            // TODO: fixme do not allow cycles
```
Comment (blair1618) [inline core/src/scene/importer.cpp]: The current implementation seems to do Very Bad Things when there is an import cycle. There's clearly not one 'correct' way to resolve a cycle, but we definitely can't crash. 

When `topologicalSort` encounters a cycle, it treats it as an error in the inputs and returns an empty list. You can check for this result and respond appropriately - for instance, using only the root scene and logging an error.

Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Yeah ... I did notice that and put in a TODO to think more on the solution.
We should not probably load/try fix the cycle right? Just do not load the scene if a cycle is found or maybe just load the default scene with no imports when a cycle is found.,

File: core/src/scene/importer.cpp [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+    if (Node styles = root["styles"]) {
+
+        for (auto styleNode : styles) {
+
+            auto style = styleNode.second;
+            //style->texture
+            if (Node texture = style["texture"]) {
+                if (texture.IsScalar()) {
+                    setNormalizedTexture(texture, {texture.Scalar()}, parentPath);
```
Comment (blair1618) [inline core/src/scene/importer.cpp]: Ok, this is gonna be somewhat tricky... the values being replaced from here down in this function, none of them are strictly URLs. They can also refer to named textures declared in the scene file.

Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Good you pointed this out. I missed it before.Will handle this in following commits. Thanks

File: core/src/scene/importer.cpp (hunk 5 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Fixed by 5d98937

File: core/src/scene/importer.cpp (hunk 4 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Fixed by 5d98937

File: core/src/scene/importer.cpp (hunk 6 above)
Comment (tallytalwar) [inline core/src/scene/importer.cpp]: Fixed by 359eed1ca3a2af641b06e1114ca3371224e95990

**Comment by tallytalwar** [comment]
Closing in favor of #776.
